Enable searching for products by SKU

jaclync opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #28508

#### Why we cannot use `sku` parameter

Searching products by SKU has been requested in Woo mobile apps in the past (ref p7bje6-2GK-p2), and I'm giving it a try to implement this for 2022 Q1 HACK Week. Currenctly, we have a `sku` parameter, but it has some issues for the use case in mobile:

- Only products with full SKU match are returned. This is used for validating SKU in the product inventory settings in mobile, and we most likley want to keep this behavior
- `search` and `sku` conditions are joined by `AND`, thus only products with the query in both product name/content/excerpt AND sku are returned
- If we make separate requests to `/wc/v3/products?search=woo` and `/wc/v3/products?sku=woo`, the total results might not match the requested page size and requires some custom handling to support pagination

#### Proposed solution

This PR attempted to support SKU search by introducing a new parameter `search_includes_sku: bool`. When `search_includes_sku` parameter is `true`, `wc_product_meta_lookup` table is joined and a `WHERE` condition is added to also (`OR`) find products with partial SKU match in the lookup table.

It seems like in core (Products tab), partial SKU match is already included in each search query (`s`):

https://github.com/woocommerce/woocommerce/blob/aa8b7a4a6b7d21ccddaac360766c4daa396b9a98/plugins%2Fwoocommerce%2Fincludes%2Fdata-stores%2Fclass-wc-product-data-store-cpt.php#L1651

We could also include the SKU search in the existing `search` parameter, but I thought it's safer to not affect pre-existing search usage given the performance impact.

### How to test the changes in this Pull Request:

Prerequisites: the test site has at least one product with a non-empty SKU (e.g. `woo-tshirt`) that isn't also in the product name of the same product

1. Run `{{site_url}}/wp-json/wc/v3/products?search={{sku_substring}}` (e.g. `search=Woo`) --> the product in the prequisite should not be returned, only products with matched name or other core product attributes are returned
2. Run the request in the previous step with `search_includes_sku=true` --> the product in the prequisite should be returned, along with other products returned in the previous step
3. Feel free to try other SKU substrings and letter casing in the search query (e.g. `tshirt`, `Tshirt`, `Woo-tshirt`)

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

jaclync committed:
Include sku partial match for products search.

jaclync committed:
Add a new parameter `search_includes_sku` so that SKU search isn't included by default.

joshbetz commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -70,6 +72,67 @@ protected function get_images( $product ) {
+			$conditions  = array(
+				$wpdb->prepare( 'wc_product_meta_lookup.sku LIKE %s', $like_search ),
+			);
+			$where      .= ' OR (' . implode( ' OR ', $conditions ) . ')';
```
Minor: We can probably simplify this a bit since it's an array with only 1 element?

joshbetz commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -70,6 +72,67 @@ protected function get_images( $product ) {
+		$search = $wp_query->get( self::$wp_query_search_param );
+		if ( ! empty( $search ) ) {
+			$join .= " LEFT JOIN {$wpdb->wc_product_meta_lookup} wc_product_meta_lookup
+						ON $wpdb->posts.ID = wc_product_meta_lookup.product_id ";
```
```suggestion
						ON {$wpdb->posts}.ID = wc_product_meta_lookup.product_id ";
```

joshbetz commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -70,6 +72,67 @@ protected function get_images( $product ) {
+		global $wpdb;
+		$search = $wp_query->get( self::$wp_query_search_param );
+		if ( ! empty( $search ) ) {
+			$join .= " LEFT JOIN {$wpdb->wc_product_meta_lookup} wc_product_meta_lookup
```
Should this just be `LEFT JOIN {$wpdb->wc_product_meta_lookup}`?

coreymckrill replied:
I suspect this is copied from [here](https://github.com/woocommerce/woocommerce/blob/cfa828d70d39b2659d3621ce487b1022cd0253f0/plugins/woocommerce/includes/class-wc-query.php#L721-L734), which also specifies the alias. Probably not necessary, but then on the next line we'd need to use `$wpdb->wc_product_meta_lookup.product_id`.

Could that method be changed to public? Then we could avoid duplicating its code here.

coreymckrill replied:
> Could that method be changed to public?

Either that, or perhaps this logic should get added in the WC_Query class instead of this REST controller...

coreymckrill commented:
> I'm also wondering if search_includes_sku is the best way to implement this. It feels a bit awkward to me to have search overloaded like this.

@joshbetz can you say more about this? Is it just that it's unconventional to have a parameter whose sole purpose is to modify the behavior of another parameter?

joshbetz commented:
@coreymckrill Having the `search` parameter function differently depending on `search_includes_sku` is the part that feels a bit weird to me. If we introduce something like `sku_search`, we can avoid overloading the existing parameter.

coreymckrill committed:
Refactor to use search_sku parameter

Instead of overloading the default search parameter.

This actually added a bit of complexity because it introduced the
possibility that a request could have both search and search_sku
parameters, and they might not be the same value.

I decided to handle the two cases differently: if both search strings
match, the assumption is that we're looking for a string in either the
content *or* the SKU field. However, if the the search strings are
different, the assumption is that we're looking for products with a
particular string in the content *and* a separate particular string
in the SKU.

New unit tests for this are still pending.

coreymckrill commented:
@joshbetz I'm still working on the unit tests, but I'm curious what you think of the approach in 6fc5838.

joshbetz commented:
> I decided to handle the two cases differently: if both search strings
match, the assumption is that we're looking for a string in either the
content *or* the SKU field. However, if the the search strings are
different, the assumption is that we're looking for products with a
particular string in the content *and* a separate particular string
in the SKU.

What do you think about treating both cases as AND?

coreymckrill commented:
> What do you think about treating both cases as AND?

Good question. I think if we treated the case of having the same value in `search` and `search_sku` as an `AND` relationship, we wouldn't get the desired outcome from #28508:

> We’d like to also support searching for a SKU in the future. So for instance, if a user enters “woo”, we should ideally fetch all products **with name or SKU** value that include the term “woo”.

But also, using `OR` in that instance better matches the current search behavior in WP Core. Currently when you do a search with the `s` parameter (or `search` in the API), the resulting SQL looks like this:

```sql
WHERE 1=1
AND (
        (
            (wp_posts.post_title LIKE '%foo%')
            OR (wp_posts.post_excerpt LIKE '%foo%')
            OR (wp_posts.post_content LIKE '%foo%')
        )
    )
AND ...
```

So the idea of making the extra clause an `OR` and inserting it via the `posts_search` filter hook instead of the more general `posts_where` is that we're really just adding another field to the group that's already being searched. (Although the code I have in place now doesn't quite do that, it's not nested with the other OR clauses. I'm working on a change...)

coreymckrill committed:
Ensure where clause gets nested correctly when terms are the same

coreymckrill commented:
With e070e21, the SQL will look like this when `search` and `search_sku` are the same:

```sql
AND (
        (
            (wp_posts.post_title LIKE '%foo%')
            OR (wp_posts.post_excerpt LIKE '%foo%')
            OR (wp_posts.post_content LIKE '%foo%')
            OR (wc_product_meta_lookup.sku LIKE '%foo%')
        )
    )
```

and this like when they're different:

```sql
AND (
        (
            (wp_posts.post_title LIKE '%foo%')
            OR (wp_posts.post_excerpt LIKE '%foo%')
            OR (wp_posts.post_content LIKE '%foo%')
        )
    )
AND (wc_product_meta_lookup.sku LIKE '%bar%')
```

(Or without the nested part at all if there is no `search` value.)

joshbetz commented:
I'm a little worried about implicitly changing the behavior of these filters based on how they're used. It seems somewhat prone to confusion.

> But also, using OR in that instance better matches the current search behavior in WP Core.

I don't know if this comparison works. Core uses OR to search multiple fields from a single string, right? In this case we're talking about potentially searching for multiple strings in different fields.

> We’d like to also support searching for a SKU in the future. So for instance, if a user enters “woo”, we should ideally fetch all products with name or SKU value that include the term “woo”.

I'll cc @anitaa1990 and @jaclync here to clarify. My understanding is that we will somehow know that it's a sku search, so the primary use case will just be to search the sku rather than both fields. (I think when this issue was originally written, there was an assumption that we would use the single `search` field to also look at the sku, thereby forcing it to be an OR.) To me, it feels more natural to further filter the results when they're both used rather than widen the search. And if we're doing that, I think the behavior should be consistent regardless of whether they match.

joshbetz commented:
Also cc @joe-keenan. My understanding is that we will somehow specify that we're searching for a sku. Does that seem OK?

joe-keenan commented:
Hi Josh and Corey, I’m just adding a link here to the discussion we’ve got going in Slack about this: p1650402426381889-slack-CGPNUU63E

coreymckrill committed:
Don't combine content and SKU search

We decided the performance hit was too great. So with this, you can
still use both the search and search_sku parameters at the same time
if you want, but the query will use `AND` so that there must be a
partial match from both fields rather than either field. To just search
SKUs, all that you need is the search_sku parameter.

coreymckrill committed:
Update tests

coreymckrill committed:
Add changelog file

coreymckrill committed:
Remove outdated code comment

coreymckrill commented:
🤔 Not sure why the changelogger check is failing, I added the file...

joshbetz commented:
Rebased against trunk. We were missing some monorepo scripts in this branch.

coreymckrill committed:
phpcs cleanup
